packages/connect/src/base/stun/tcp/client.ts

```diff
@@ -0,0 +1,439 @@
+import { type Socket, createConnection } from 'net'
+
+import { decode, constants, createMessage, createTransaction } from 'stun'
+
+// @ts-ignore untyped module
+import retimer from 'retimer'
+
+import { Multiaddr } from '@multiformats/multiaddr'
+
+import { u8aToHex, u8aAddrToString, u8aToNumber, u8aEquals } from '@hoprnet/hopr-utils'
+import { STUN_TCP_TIMEOUT } from './constants.js'
+import { CODE_IP4, CODE_IP6, CODE_DNS4, CODE_DNS6 } from '../../../constants.js'
+import {
+  isStunErrorResponse,
+  isStunRequest,
+  isStunSuccessResponse,
+  kStunTypeMask,
+  STUN_EXPOSED_CHECK_RESPOSE,
+  STUN_QUERY_STATE
+} from '../constants.js'
+
+import debug from 'debug'
+
+const log = debug('hopr-connect:stun:tcp')
+const error = debug('hopr-connect:stun:tcp:error')
+const verbose = debug(`hopr-connect:verbose:stun:tcp`)
+
+type Interface = {
+  family: 'IPv4' | 'IPv6'
+  port: number
+  address: string
+}
+
+type Request = {
+  multiaddr: Multiaddr
+  responsePort?: number
+  response?: Interface
+  timeout: any
+  state: STUN_QUERY_STATE
+}
+
+type Requests = Map<string, Request>
+
+/**
+ * Encodes STUN message and sends it to the STUN server using
+ * a secondary TCP connection.
+ *
+ * @param multiaddr address to contact
+ * @param tId transactionId, used to distinguish requests
+ * @param responsePort port on which the response is expected
+ * @param signal used to terminate request
+ * @param onUpdate callback, called on incoming responses
+ */
+function createRequest(
+  multiaddr: Multiaddr,
+  tId: Buffer,
+  responsePort: number | undefined,
+  signal: AbortSignal,
+  onUpdate: (response: { response?: Interface; transactionId: Buffer }) => void
+): void {
+  const tuples = multiaddr.tuples()
+
+  if (tuples.length == 0) {
+    throw Error(`Cannot perform STUN request: empty Multiaddr`)
+  }
+
+  let address: string
+
+  switch (tuples[0][0]) {
+    case CODE_DNS4:
+    case CODE_DNS6:
+      address = new TextDecoder().decode(tuples[0][1]?.slice(1) as Uint8Array)
+      break
+    case CODE_IP6:
+      address = u8aAddrToString(tuples[0][1] as Uint8Array, 'IPv6')
+      break
+    case CODE_IP4:
+      address = `::ffff:${u8aAddrToString(tuples[0][1] as Uint8Array, 'IPv4')}`
+      break
+    default:
+      throw Error(`Invalid address: ${multiaddr.toString()}`)
+  }
+
+  const port: number | undefined = tuples.length >= 2 ? u8aToNumber(tuples[1][1] as Uint8Array) : undefined
+
+  const message = createMessage(constants.STUN_BINDING_REQUEST, tId)
+
+  // Response port can be 0
+  if (responsePort != undefined) {
+    // Ask the STUN server to reply on a specific port, see RFC 5780
+    message.addAttribute(constants.STUN_ATTR_RESPONSE_PORT, responsePort)
+  }
```

Comment:
Can we separate this to some function so we prevent duplication in `udp/client.ts` along lines 178-208 ?

Reply:
done, thx for the suggestion